lib/charms/mlops_libs/v0/k8s_service_info.py

```diff
@@ -0,0 +1,312 @@
+
+    Args:
+        requirer_charm (CharmBase): the requirer application
+        relation_name (str, optional): the name of the relation
+
+    Attributes:
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    def __init__(self, requirer_charm, relation_name: Optional[str] = DEFAULT_RELATION_NAME):
+        super().__init__(requirer_charm, relation_name)
+        self.relation_name = relation_name
+
+    @staticmethod
+    def _validate_relation(relation: Relation) -> None:
+        """Series of checks for the relation and relation data.
+
+        Args:
+            relation (Relation): the relation object to run the checks on
+
+        Raises:
+            KubernetesServiceInfoRelationDataMissingError if data is missing or incomplete
+            KubernetesServiceInfoRelationMissingError: if there is no related application
+            ops.model.TooManyRelatedAppsError: if there is more than one related application
+        """
+        # Raise if there is no related application
+        if not relation:
+            raise KubernetesServiceInfoRelationMissingError()
+
+        # Extract remote app information from relation
+        remote_app = relation.app
+        # Get relation data from remote app
+        relation_data = relation.data[remote_app]
+
+        # Raise if there is no data found in the relation data bag
+        if not relation_data:
+            raise KubernetesServiceInfoRelationDataMissingError(
+                f"No data found in relation {relation.name} data bag."
+            )
+
+        # Check if the relation data contains the expected attributes
+        missing_attributes = [
+            attribute for attribute in REQUIRED_ATTRIBUTES if attribute not in relation_data
+        ]
+        if missing_attributes:
+            raise KubernetesServiceInfoRelationDataMissingError(
+                f"Missing attributes: {missing_attributes} in relation {relation.name}"
+            )
+
+    def get_data(self) -> KubernetesServiceInfoObject:
+        """Return a KubernetesServiceInfoObject containing Kubernetes Service information.
+
+        Raises:
+            KubernetesServiceInfoRelationDataMissingError: if data is missing entirely or some attributes
+            KubernetesServiceInfoRelationMissingError: if there is no related application
+        """
+        # Validate relation data
+        # Raises TooManyRelatedAppsError if related to more than one app
+        relation = self.model.get_relation(self.relation_name)
+        self._validate_relation(relation=relation)
+
+        # Get relation data from remote app
+        relation_data = relation.data[relation.app]
+
+        return KubernetesServiceInfoObject(
+            name=relation_data["svc_name"], port=relation_data["svc_port"]
+        )
+
+
+class KubernetesServiceInfoProvider(Object):
+    """Base class that represents the provider end of the Kubernetes Service info relation.
```

Comment:
Could we be more direct here?  When comparing `KubernetesServiceInfoProvider` and KubernetesServiceInfoRequirer` I realised I misunderstood these classes and think the docstring could help keep things clearer.  They both have almost the same docstring, but are completely different in function.

"represents the provider end" could mean so much - we could be more specific.  And the docstring doesn't mention how this will observe some events, or what it'll do on those events.  

Maybe: Implements the Provider end of the Kubernetes Service Info relation, observing relation events to send data to related applications

(although I think that'll be too long).  And then elaborate a bit more on the following lines